ui/revenuecatui/src/main/kotlin/com/revenuecat/purchases/ui/revenuecatui/helpers/OfferingToStateMapper.kt

```diff
@@ -1,28 +1,116 @@
+import com.revenuecat.purchases.paywalls.PaywallData
 import com.revenuecat.purchases.ui.revenuecatui.PaywallViewMode
 import com.revenuecat.purchases.ui.revenuecatui.data.PaywallViewState
+import com.revenuecat.purchases.ui.revenuecatui.data.processed.PaywallTemplate
 import com.revenuecat.purchases.ui.revenuecatui.data.processed.TemplateConfigurationFactory
 import com.revenuecat.purchases.ui.revenuecatui.data.processed.VariableDataProvider
+import com.revenuecat.purchases.ui.revenuecatui.data.processed.VariableProcessor
+import com.revenuecat.purchases.ui.revenuecatui.errors.PaywallValidationError
+import com.revenuecat.purchases.ui.revenuecatui.extensions.createDefault
+import com.revenuecat.purchases.ui.revenuecatui.extensions.defaultTemplate
+import com.revenuecat.purchases.ui.revenuecatui.utils.Result
 
-@Suppress("ReturnCount", "TooGenericExceptionCaught")
-internal fun Offering.toPaywallViewState(variableDataProvider: VariableDataProvider): PaywallViewState {
+internal data class ValidationResult(
+    val displayablePaywall: PaywallData,
+    val template: PaywallTemplate,
+    val error: PaywallValidationError?,
+)
+
+internal sealed class PaywallValidationResult {
+    data class Success(val displayablePaywall: PaywallData, val template: PaywallTemplate) : PaywallValidationResult()
+    data class Error(
+        val displayablePaywall: PaywallData,
+        val template: PaywallTemplate,
+        val error: PaywallValidationError,
+    ) : PaywallValidationResult()
+}
+
+internal fun Offering.validatedPaywall(
+    packageName: String,
+    currentColors: Colors,
+): ValidationResult {
     val paywallData = this.paywall
-        ?: return PaywallViewState.Error("No paywall data for offering: $identifier")
-    return try {
-        val templateConfiguration = TemplateConfigurationFactory.create(
-            variableDataProvider = variableDataProvider,
-            mode = PaywallViewMode.FULL_SCREEN,
-            paywallData = paywallData,
-            packages = availablePackages,
-            activelySubscribedProductIdentifiers = emptySet(), // TODO-PAYWALLS: Check for active subscriptions
+        ?: return ValidationResult(
+            PaywallData.createDefault(packages = availablePackages, packageName, currentColors),
+            PaywallData.defaultTemplate,
+            PaywallValidationError.MissingPaywall,
+        )
+
+    return when (val result = paywallData.validate()) {
+        is Result.Error -> ValidationResult(
+            PaywallData.createDefault(packages = availablePackages, packageName, currentColors),
+            PaywallData.defaultTemplate,
+            result.value,
         )
-        PaywallViewState.Loaded(
-            templateConfiguration = templateConfiguration,
-            selectedPackage = templateConfiguration.packages.default,
+        is Result.Success -> ValidationResult(
+            paywallData,
+            result.value,
+            null,
         )
-    } catch (e: Exception) {
```

Comment:
Since we're not doing this anymore, it may crash in the `TemplateConfigurationFactory.create` since we `require` a that there is at least some packages. Should we perform that validation before to avoid the crash?

Reply:
Right, I missed those `require`

Reply:
[`ec1a4f4` (#1273)](https://github.com/RevenueCat/purchases-android/pull/1273/commits/ec1a4f4627cd38195ec4d60ef925d5e768aa2968) and [`6bb5ebc` (#1273)](https://github.com/RevenueCat/purchases-android/pull/1273/commits/6bb5ebcfee6132a6cd4fea19abd12ec30567d482)